src/mon/OSDMonitor.cc

```diff
@@ -3481,14 +3499,69 @@ void OSDMonitor::get_removed_snaps_range(
 
 int OSDMonitor::get_version(version_t ver, bufferlist& bl)
 {
-    if (inc_osd_cache.lookup(ver, &bl)) {
-      return 0;
-    }
-    int ret = PaxosService::get_version(ver, bl);
-    if (!ret) {
-      inc_osd_cache.add(ver, bl);
+  return get_version(ver, mon->get_quorum_con_features(), bl);
+}
+
+void OSDMonitor::reencode_incremental_map(bufferlist& bl, uint64_t features)
+{
+  OSDMap::Incremental inc;
+  bufferlist::iterator q = bl.begin();
+  inc.decode(q);
+  // always encode with subset of osdmap's canonical features
+  uint64_t f = features & inc.encode_features;
+  dout(20) << __func__ << " " << inc.epoch << " with features " << f
+	   << dendl;
+  bl.clear();
+  if (inc.fullmap.length()) {
+    // embedded full map?
+    OSDMap m;
+    m.decode(inc.fullmap);
+    inc.fullmap.clear();
+    m.encode(inc.fullmap, f | CEPH_FEATURE_RESERVED);
+  }
+  if (inc.crush.length()) {
+    // embedded crush map
+    CrushWrapper c;
+    auto p = inc.crush.begin();
+    c.decode(p);
+    inc.crush.clear();
+    c.encode(inc.crush, f);
+  }
+  inc.encode(bl, f | CEPH_FEATURE_RESERVED);
+}
+
+void OSDMonitor::reencode_full_map(bufferlist& bl, uint64_t features)
+{
+  OSDMap m;
+  bufferlist::iterator q = bl.begin();
+  m.decode(q);
+  // always encode with subset of osdmap's canonical features
+  uint64_t f = features & m.get_encoding_features();
+  dout(20) << __func__ << " " << m.get_epoch() << " with features " << f
+	   << dendl;
+  bl.clear();
+  m.encode(bl, f | CEPH_FEATURE_RESERVED);
+}
+
+int OSDMonitor::get_version(version_t ver, uint64_t features, bufferlist& bl)
+{
+  uint64_t significant_features = OSDMap::get_significant_features(features);
+  if (inc_osd_cache.lookup({ver, significant_features}, &bl)) {
+    return 0;
+  }
+  int ret = PaxosService::get_version(ver, bl);
+  if (!ret) {
```

Comment:
i agree. we should *always* reencode the bl retrieved from the underlying monstore. so, https://github.com/ceph/ceph/pull/21605/commits/8b5fb529ed5e19ce3a1cf48335b6219e66da603c#diff-0a5db46a44ae9900e226289a810f10e8R3553 should be 
```c++
if (!ret) {
  return ret;
}
if (siginificant_features != 
    OSDMap::get_significant_features(mon->get_quorum_con_features())) {
  reencode_incremental_map(bl, features);
}
inc_osd_cache.add({ver, significant_features}, bl);
// ...
```

Reply:
It is exactly how it works now,  but yes, your code looks more strightforward

Reply:
@xiaoxichen please re-read the code. they are different. i am not proposing a cosmetic change.

Reply:
fixed

Reply:
ahh, sorry. i was wrong. 
```c++
if (!ret) {
  // re-encode
}
return ret;
```
is correct.

Reply:
sorry, i was wrong. it was me who mis-read the code.

should be 
```c++
if (ret) {
  return ret;
}
// do the re-encoding
return 0;
```